core/src/net/tcp_listener.cpp

```diff
@@ -0,0 +1,115 @@
+#ifdef _WIN32
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#pragma comment(lib, "Ws2_32.lib")
+#else
+#include <arpa/inet.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#endif
+
+#include <cstring>
+#include <memory>
+
+#include <cubos/core/log.hpp>
+#include <cubos/core/net/tcp_listener.hpp>
+#include <cubos/core/net/tcp_stream.hpp>
+#include <cubos/core/reflection/external/cstring.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+
+using cubos::core::net::Address;
+using cubos::core::net::TcpListener;
+using cubos::core::net::TcpStream;
+
+/// @brief Cross-platform utility to get last error message.
+/// @return Message.
+/// @todo Move this to a higher module for general purposes?
+static std::string error()
+{
+#ifdef _WIN32
+    static char message[256] = {0};
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, 0, WSAGetLastError(), 0, message, 256, 0);
+    char* nl = strrchr(message, '\n');
+    if (nl)
+        *nl = 0;
+    return {message};
+#else
+    return {std::strerror(errno)};
+#endif
+}
+
+TcpListener::TcpListener()
+{
+#ifdef _WIN32
+    WSADATA wsa;
+    CUBOS_ASSERT(WSAStartup(MAKEWORD(2, 2), &wsa) == 0, " WSAStartup failed: {}", WSAGetLastError());
+#endif
+}
+
+TcpListener::~TcpListener()
+{
+    close();
+}
+
+bool TcpListener::listen(const Address& address, uint16_t port, int maxQueueLength)
+{
+    mSock = socket(AF_INET, SOCK_STREAM, 0);
+    if (mSock == -1)
+    {
+        CUBOS_ERROR("Failed to allocate TCP socket: {}", error());
+        return false;
+    }
+
+    sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(port);
+    addr.sin_addr.s_addr = inet_addr(address.toString().c_str());
+
+    if (::bind(mSock, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) == -1)
+    {
+        CUBOS_ERROR("Failed to bind TCP socket: {}", error());
+        return false;
+    }
+
+    if (::listen(mSock, maxQueueLength) < 0)
+    {
+        CUBOS_ERROR("Failed to listen to TCP socket: {}", error());
+        return false;
+    }
+
+    CUBOS_INFO("TCP listener socket '{}' at address {} with port '{}'", mSock, address.toString(), port);
+    return true;
+}
+
+bool TcpListener::accept(TcpStream& stream)
+{
+    int clientSocket = ::accept(mSock, nullptr, nullptr);
+    if (clientSocket == INVALID_SOCKET)
+    {
+        CUBOS_ERROR("Failed to accept TCP connection: {}", error());
+        return false;
+    }
+
+    CUBOS_INFO("Connecting stream to socket");
+    stream.connect(clientSocket);
```

Comment:
H'm I don't like this 'connect' wording, because you're not connecting the stream to a socket, you're just replacing the socket of the stream with a connected one, right?
I think you could just add a `.inner(socket)` method which replaces the inner socket with the given one, destroying the old one.